### PR TITLE
[SPARK-8531] [ML] Update ML user guide for MinMaxScaler

### PR DESCRIPTION
**Pull request by hhbyyh**
jira: https://issues.apache.org/jira/browse/SPARK-8531

Update ML user guide for MinMaxScaler

### CODE REVIEW

File: docs/ml-features.md [hunk 1]
```diff
@@ -865,6 +865,7 @@ val scaledData = scalerModel.transform(dataFrame)
 {% highlight java %}
 import org.apache.spark.api.java.JavaRDD;
 import org.apache.spark.ml.feature.StandardScaler;
+import org.apache.spark.ml.feature.StandardScalerModel;
```
Comment (jkbradley) [inline docs/ml-features.md]: good catch!

File: docs/ml-features.md [hunk 2]
```diff
@@ -905,6 +906,74 @@ scaledData = scalerModel.transform(dataFrame)
...
+`MinMaxScaler` transforms a dataset of `Vector` rows, rescaling each feature to a specific range (often [0, 1]).  It takes parameters:
+
+* `min`: 0.0 by default. Lower bound after transformation, shared by all features.
+* `max`: 1.0 by default. Upper bound after transformation, shared by all features.
+
+`MinMaxScaler` computes summary statistics on a data set and produces a `MinMaxScalerModel`. The model can then transform each feature individually such that it is in the given range.
+
+The rescaled value for a feature E is calculated as,
+
+  Rescaled(e_i) = \frac{e_i - E_{min}}{E_{max} - E_{min}} * (max - min) + min
```
Comment (jkbradley) [inline docs/ml-features.md]: Could you please make this render as Latex?  You can follow the examples in mllib-linear-methods.md

Same for other equations below.

File: docs/ml-features.md [hunk 3]
```diff
@@ -905,6 +906,74 @@ scaledData = scalerModel.transform(dataFrame)
...
+
+For the case E_{max} == E_{min}, Rescaled(e_i) = 0.5 * (max + min)
+
+Note that since zero values will probably be transformed to non-zero values, output of the transformer will be DenseVector even for sparse input.
+
+More details can be found in the API docs for
+[MinMaxScaler](api/scala/index.html#org.apache.spark.ml.feature.MinMaxScaler) and
```
Comment (jkbradley) [inline docs/ml-features.md]: Could you please put this in the Scala code tab like this:
```
<div data-lang="scala">
(you can put it here)
{% highlight scala %}
```
And then could you also please add a reference to the other APIs under those code tabs? Thanks!  (I'm trying to follow this pattern nowadays.)

File: docs/ml-features.md [hunk 4]
```diff
@@ -905,6 +906,74 @@ scaledData = scalerModel.transform(dataFrame)
...
+import org.apache.spark.ml.feature.MinMaxScaler
+import org.apache.spark.mllib.util.MLUtils
+
+val data = MLUtils.loadLibSVMFile(sc, "data/mllib/sample_libsvm_data.txt")
+val dataFrame = sqlContext.createDataFrame(data)
+val scaler = new MinMaxScaler()
+  .setInputCol("features")
+  .setOutputCol("scaledFeatures")
+
+// Compute summary statistics by fitting the StandardScaler
```
Comment (jkbradley) [inline docs/ml-features.md]: Please update the inline comments.

File: docs/ml-features.md [hunk 5]
```diff
@@ -905,6 +906,74 @@ scaledData = scalerModel.transform(dataFrame)
...
+val scalerModel = scaler.fit(dataFrame)
+
+// Normalize each feature to have unit standard deviation.
```
Comment (jkbradley) [inline docs/ml-features.md]: Please update the inline comments.

File: docs/ml-features.md [hunk 6]
```diff
@@ -905,6 +906,74 @@ scaledData = scalerModel.transform(dataFrame)
...
+import org.apache.spark.sql.DataFrame;
+
+JavaRDD<LabeledPoint> data =
+  MLUtils.loadLibSVMFile(jsc.sc(), "data/mllib/sample_libsvm_data.txt").toJavaRDD();
+DataFrame dataFrame = jsql.createDataFrame(data, LabeledPoint.class);
+MinMaxScaler scaler = new MinMaxScaler()
+  .setInputCol("features")
+  .setOutputCol("scaledFeatures");
+
+// Compute summary statistics by fitting the StandardScaler
```
Comment (jkbradley) [inline docs/ml-features.md]: Please update the inline comments.

File: docs/ml-features.md [hunk 7]
```diff
@@ -905,6 +906,74 @@ scaledData = scalerModel.transform(dataFrame)
...
+MinMaxScalerModel scalerModel = scaler.fit(dataFrame);
+
+// Normalize each feature to have unit standard deviation.
```
Comment (jkbradley) [inline docs/ml-features.md]: Please update the inline comments.

**Comment by jkbradley** [comment]
Would you mind adding an example for Python too?  Or we can make that a separate PR if you prefer.

File: docs/ml-features.md (hunk 2 above)
Comment (hhbyyh) [inline docs/ml-features.md]: Sure, do you mean adding `\begin{equation} ....\end{equation}`, or I should further modify the equation.

**Comment by hhbyyh** [comment]
@jkbradley Thanks for the review. 
I'm not sure if the latex part looks good.
And for python document, since python interface for MinMaxScaler is still under review, so I didn't add the example.

File: docs/ml-features.md [hunk 8]
```diff
@@ -1098,6 +1099,76 @@ scaledData = scalerModel.transform(dataFrame)
...
+  Rescaled(e_i) = \frac{e_i - E_{min}}{E_{max} - E_{min}} * (max - min) + min
+\end{equation}`
+For the case `E_{max} == E_{min}`, `Rescaled(e_i) = 0.5 * (max + min)`
+
+Note that since zero values will probably be transformed to non-zero values, output of the transformer will be DenseVector even for sparse input.
+
+The following example demonstrates how to load a dataset in libsvm format and then rescale each feature to [0, 1].
+
+<div class="codetabs">
+<div data-lang="scala">
```
Comment (jkbradley) [inline docs/ml-features.md]: The links are not generated correctly, but you can fix it by modifying this line:
```
<div data-lang="scala" markdown="1">
```

File: docs/ml-features.md [hunk 9]
```diff
@@ -1098,6 +1099,76 @@ scaledData = scalerModel.transform(dataFrame)
...
+
+// Compute summary statistics and generate MinMaxScalerModel
+val scalerModel = scaler.fit(dataFrame)
+
+// rescale each feature to range [min, max].
+val scaledData = scalerModel.transform(dataFrame)
+{% endhighlight %}
+</div>
+
+<div data-lang="java">
```
Comment (jkbradley) [inline docs/ml-features.md]: same here (add ```markdown="1"```)

File: docs/ml-features.md [hunk 10]
```diff
@@ -1098,6 +1099,76 @@ scaledData = scalerModel.transform(dataFrame)
...
+More details can be found in the API docs for
+[MinMaxScaler](api/java/index.html#org.apache.spark.ml.feature.MinMaxScaler) and
```
Comment (jkbradley) [inline docs/ml-features.md]: Link should be:
```[api/java/org/apache/spark/ml/feature/MinMaxScaler.html]```
(similarly for model)

**Comment by jkbradley** [comment]
The latex part looks much better now.

**Comment by jkbradley** [comment]
ping!

**Comment by hhbyyh** [comment]
Sorry for the delay. I am in my hometown and has poor internet connection. I'll be back to office tomorrow and start catching up.

**Comment by jkbradley** [comment]
Thank you for updating it.  LGTM.
Merging with master and branch-1.5

**Pull request closed without merging**